gh-132657: Add maybe_enable_deferred_ref_count().

nascheme opened the pull request:
If we are specializing to LOAD_GLOBAL_MODULE or LOAD_ATTR_MODULE, try to enable deferred reference counting for the value, if the object is owned by a different thread.

Add a benchmark to `ftscalingbench.py` that shows the benefit of this change.  The `deepcopy` benchmark uses globals like `_local_atomic_types`.

<!-- gh-issue-number: gh-132657 -->
* Issue: gh-132657
<!-- /gh-issue-number -->

nascheme committed:
Add maybe_enable_deferred_ref_count().

If we are specializing to LOAD_GLOBAL_MODULE, set deferred reference
counting for the value, if it meets criteria. For now, it's only done
for frozenset objects.

nascheme committed:
Improve maybe_enable_deferred_ref_count() function.

Use it for LOAD_ATTR_MODULE in addition to LOAD_GLOBAL_MODULE.  Don't
enable deferred ref counts if the object is owned by the current thread.
Specialized bytecode is per-thread so this works.  Enable for
frozensets, tuples and type objects.

colesbury commented on Python/specialize.c:
```diff
@@ -355,6 +355,28 @@ static int function_kind(PyCodeObject *code);
+    PyObject *op;
+    if (PyDict_GetItemRef(dict, name, &op) != 1) {
+        return;
+    }
```
Let's combine this lookup with the earlier `_PyDict_LookupIndex`. `_PyDict_LookupIndex` already gets the value internally and throws it away.

colesbury commented on Python/specialize.c:
```diff
@@ -355,6 +355,28 @@ static int function_kind(PyCodeObject *code);
+            _PyObject_HasDeferredRefcount(op)) {
+        Py_DECREF(op);
+        return;
+    }
```
The `!PyType_IS_GC` and `_PyObject_HasDeferredRefcount` checks aren't really necessary because `PyUnstable_Object_EnableDeferredRefcount` handles those cases internally.

colesbury commented on Python/specialize.c:
```diff
@@ -355,6 +355,28 @@ static int function_kind(PyCodeObject *code);
+    }
+    if (PyFrozenSet_Check(op) || PyTuple_Check(op) || PyType_Check(op)) {
+        PyUnstable_Object_EnableDeferredRefcount(op);
+    }
```
I think we should do this for all objects, not just a few types. Otherwise, I think we will keep run into scaling bottlenecks with global variables.

Yhg1s replied:
Hm... That's going to push a lot of objects to the GC for deallocation in programs that do a lot of work at the global level (but do have a few functions that are called enough that use those globals.) I'm thinking of naively written "scripts". Then again, it would only do that for objects that participate in GC (i.e. not strings or integers), so it's *probably* fine. I mean, nobody's too surprised when things aren't deallocated quite as quickly as they expect... right?

Yeah, I think this should be fine for main. I definitely wouldn't backport it to 3.14.

colesbury replied:
Definitely agree regarding not backporting to 3.14.

I think the above check that excludes owned by the current thread should help with most scripts -- it'll only affect global variables accessed by multiple threads.

Another thing that may help is that I think the `PyUnstable_Object_EnableDeferredRefcount` call only happens at specialization time, not every time a variable is updated or accessed.

nascheme replied:
Yeah, I'm a bit worried about increased memory usage.  Since it's only for objects shared between threads and only for hot LOAD_GLOBAL/LOAD_ATTR instructions, maybe it's okay.  Perhaps we should consider changing `gc_should_collect()` to check not only the young object count but also the process memory size increase as well.  The condition with `count < gcstate->long_lived_total / 4` could be done with an "or" condition with the process size check.

colesbury reviewed (commented):
I think this makes sense and is worth doing. Let's also add a scaling test to ftscalingbench.py

nascheme committed:
Enable deferred counting for more types.

Also, avoid PyDict_GetItemRef() call by returning the value when the
index is looked up (as previously discarded).

nascheme committed:
Add deepcopy benchmark to ftscalingbench.

This is taken from the PR GH-132658.

nascheme committed:
Add comment for maybe_enable_deferred_ref_count().

nascheme committed:
Merge remote-tracking branch 'origin/main' into specialize-enable-deferred

nascheme committed:
Add NEWS.

nascheme committed:
Merge remote-tracking branch 'origin/main' into specialize-enable-deferred

nascheme committed:
Don't use :opcode: markup.

nascheme committed:
Merge 'origin/main' into specialize-enable-deferred

nascheme committed:
Merge branch 'main' into specialize-enable-deferred
